### PR TITLE
feat: popup wallet settings

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-4335
https://mask.atlassian.net/browse/MF-4604

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/icons/general/UserOutline.svg
Comment (UncleBill) [inline packages/icons/general/UserOutline.svg:1]: image

File: packages/mask/src/extension/popups/pages/Wallet/WalletSettings/Contacts.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+export function Contacts() {
+    const { t } = useI18N()
+    const theme = useTheme()
+    const wallet = useWallet()
+    const wallets = useWallets()
+    const contacts = useContacts()
+    const { classes } = useStyles()
+    const navigate = useNavigate()
+    const totalContacts = useMemo(() => wallets.length + contacts.length, [wallets.length + contacts.length])
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/pages/Wallet/WalletSettings/Contacts.tsx]:
```suggestion
    const totalContacts = wallets.length + contacts.length
```

File: packages/mask/src/extension/popups/modals/ChooseCurrencyModal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        fontSize: 12,
+        fontWeight: 700,
+        marginLeft: 8,
+    },
+}))
+
+interface NetworkItemProps {
+    fiatCurrencyType: FiatCurrencyType
+}
+const CurrencyItem = memo(function NetworkItem({ fiatCurrencyType }: NetworkItemProps) {
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/modals/ChooseCurrencyModal/index.tsx]:
```suggestion
interface CurrencyItemProps {
    fiatCurrencyType: FiatCurrencyType
}
const CurrencyItem = memo(function CurrencyItem({ fiatCurrencyType }: CurrencyItemProps) {
```

File: packages/mask/src/extension/popups/pages/Wallet/ExportPrivateKey/index.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    {value?.privateKey ? (
+                        <Typography
+                            p={1.5}
+                            style={{
+                                background: theme.palette.maskColor.input,
+                                wordWrap: 'break-word',
+                                borderRadius: 8,
+                                height: 240,
+                            }}>
+                            {value?.privateKey}
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/pages/Wallet/ExportPrivateKey/index.tsx]:
```suggestion
                            {value.privateKey}
```

File: packages/mask/src/extension/popups/pages/Wallet/ExportPrivateKey/index.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            jsonFile: await WalletRPC.exportKeyStoreJSON(wallet.address, state?.password),
+            privateKey: await WalletRPC.exportPrivateKey(wallet.address, state?.password),
+        }
+    }, [wallet, state?.password])
+
+    const [currentTab, onChange] = useTabs(TabType.PrivateKey, TabType.JsonFile)
+
+    const handleCopy = useCallback(() => {
+        if (!value?.privateKey) return
+        copyToClipboard(value?.privateKey)
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/pages/Wallet/ExportPrivateKey/index.tsx]:
```suggestion
        copyToClipboard(value.privateKey)
```

File: packages/web3-shared/evm/src/libs/PayloadEditor.ts [hunk 5]
```diff
@@ -88,6 +88,7 @@ export class PayloadEditor {
 
     private getRawConfig() {
         const { method, params } = this.payload
+        const coder = ABICoder as unknown as ABICoder.AbiCoder
```
Comment (guanbinrui) [inline packages/web3-shared/evm/src/libs/PayloadEditor.ts:91]: Make a lib here: `packages/web3-shared/evm/src/libs/AbiCoder.ts` for one-time type fixing.

Comment (guanbinrui) [inline packages/web3-shared/evm/src/libs/PayloadEditor.ts]: If you prefer using plain function calls you could re-export abi coder methods here: `/packages/web3-shared/evm/src/helpers/`.

File: packages/web3-shared/base/src/specs/index.ts [hunk 6]
```diff
@@ -917,12 +924,16 @@ export interface SettingsState extends Startable {
     allowTestnet?: Subscription<boolean>
     /** The currency of estimated values and prices. */
     currencyType?: Subscription<CurrencyType>
+    /** The fiat currency of estimated values and prices. */
+    fiatCurrencyType?: Subscription<FiatCurrencyType>
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: The `currencyType` is reserved for fiat currency as its original design. Please choose one to keep between `fiatCurrencyType` and `currencyType`.

File: packages/web3-shared/base/src/helpers/formatMarketCap.ts [hunk 7]
```diff
@@ -20,17 +20,23 @@ const boundaryValues = [
...
 
 /**
  * format Market Cap
  * @returns format result
  * @param value
  * @param currency
+ * @param currencyRate
  */
-export function formatMarketCap(value: BigNumber.Value, currency = 'USD') {
+export function formatMarketCap(value: BigNumber.Value, currency = 'USD', currencyRate = 1) {
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/helpers/formatMarketCap.ts]:
```suggestion
export function formatMarketCap(value: BigNumber.Value, currency = CurrencyType.USD, currencyRate = 1) {
```

File: packages/web3-providers/src/Web3/EVM/providers/BaseHosted.ts [hunk 8]
```diff
@@ -146,6 +146,12 @@ export class BaseHostedProvider
     }
 
     override async renameWallet(address: string, name: string) {
+        const isNameExists = this.walletStorage?.wallets.value
```
Comment (guanbinrui) [inline packages/web3-providers/src/Web3/EVM/providers/BaseHosted.ts:149]: Such a strict rule here. 🤔

File: packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    const { t } = useI18N()
+    const theme = useTheme()
+    const { classes, cx } = useStyles()
+    const { value: autoLockerTime = DEFAULT_MIN_AUTO_LOCKER_TIME } = useWalletAutoLockTime()
+
+    const [time, setTime] = useState<string>()
+
+    const initialTime = millisecondsToMinutes(autoLockerTime).toString()
+
+    const error = Number.isNaN(Number(time ?? initialTime))
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx:67]: Use `date-fns` to parse a timestamp.

File: packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    const [{ loading }, setAutoLockerTime] = useAsyncFn(async (time: number) => {
+        await WalletRPC.setAutoLockerTime(
+            time === 0 ? time : time < DEFAULT_MIN_AUTO_LOCKER_TIME ? DEFAULT_MIN_AUTO_LOCKER_TIME : time,
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx]: `Math.max()`

File: packages/mask/src/extension/popups/modals/WalletRenameModal/index.tsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+function WalletRenameDrawer({ wallet, ...rest }: WalletRenameDrawerProps) {
+    const { t } = useI18N()
+    const theme = useTheme()
+    const [name, setName] = useState('')
+    const [error, setError] = useState('')
+
+    const [{ loading }, handleClick] = useAsyncFn(async () => {
+        if (!name || !wallet) return
+        if (name.length > 20 || name.length < 3) {
+            setError('popups_wallet_settings_rename_tips')
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/WalletRenameModal/index.tsx]: Looks like an unused i18n key.

File: packages/mask/src/extension/popups/pages/Wallet/GasSetting/Prior1559GasSetting.tsx [hunk 12]
```diff
@@ -266,8 +275,8 @@ export const Prior1559GasSetting = memo(() => {
                                             formatWeiToEther(gasPrice)
                                                 .times(nativeTokenPrice)
                                                 .times(minGasLimit || 21000),
-                                            'USD',
-                                            { onlyRemainTwoDecimal: true },
+                                            fiatCurrencyType,
+                                            { onlyRemainTwoDecimal: true, fiatCurrencyRate },
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/GasSetting/Prior1559GasSetting.tsx]: Using a component to replace the pure function, tears the labor of always passing a currency type prop.

File: packages/mask/src/extension/popups/pages/Wallet/TokenDetail/TrendingChart.tsx [hunk 13]
```diff
@@ -17,10 +19,14 @@ interface TrendingChartProps extends HTMLProps<SVGSVGElement> {
 
 export function TrendingChart({ stats, ...props }: TrendingChartProps) {
     const svgRef = useRef<SVGSVGElement>(null)
-    const chartData = useMemo(() => stats.map(([date, price]) => ({ date: new Date(date), value: price })), [stats])
+    const { fiatCurrencyType, fiatCurrencyRate } = useContainer(WalletContext)
+    const chartData = useMemo(
+        () => stats.map(([date, price]) => ({ date: new Date(date), value: price * fiatCurrencyRate })),
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/TokenDetail/TrendingChart.tsx]: Currency conversion should concern with the loss of precision and overflow issues, try `multipliedBy`.

File: packages/mask/src/extension/popups/pages/Wallet/WalletSettings/AutoLock.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    const theme = useTheme()
+    const { classes } = useStyles()
+
+    const { value } = useWalletAutoLockTime()
+
+    return (
+        <ListItem
+            className={classes.item}
+            onClick={() =>
+                WalletAutoLockSettingModal.openAndWaitForClose({
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/WalletSettings/AutoLock.tsx]: Why wait for closing here?

File: packages/mask/src/extension/popups/pages/Wallet/WalletSettings/ShowPrivateKey.tsx [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+export function ShowPrivateKey() {
+    const { t } = useI18N()
+    const theme = useTheme()
+    const { classes } = useStyles()
+
+    return (
+        <ListItem
+            className={classes.item}
+            onClick={() =>
+                ShowPrivateKeyModal.openAndWaitForClose({
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/WalletSettings/ShowPrivateKey.tsx]: Not necessary for waiting here.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 16]
```diff
@@ -215,7 +215,14 @@ export async function exportPrivateKey(address: string, unverifiedPassword?: str
 
 export async function exportKeyStoreJSON(address: string, unverifiedPassword?: string) {
     if (unverifiedPassword) await password.verifyPasswordRequired(unverifiedPassword)
-    const password_ = await password.INTERNAL_getPasswordRequired()
+    let password_: string = ''
+    try {
+        password_ = await password.INTERNAL_getPasswordRequired()
+    } catch {
+        if (!unverifiedPassword) throw new Error('No password set yet or expired.')
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts]: Use `hasSecret` or `hasSafeSecret` to detect whether a password was set before.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 17]
```diff
@@ -193,7 +193,7 @@ export async function exportMnemonicWords(address: string, unverifiedPassword?:
 
 export async function exportPrivateKey(address: string, unverifiedPassword?: string) {
     if (unverifiedPassword) await password.verifyPasswordRequired(unverifiedPassword)
-    const password_ = await password.INTERNAL_getPasswordRequired()
+    const password_ = (await password.INTERNAL_getPasswordRequired()) ?? unverifiedPassword
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts]: The user won't know the mastery password. It cannot be treated as a password here.

File: packages/web3-shared/base/src/helpers/formatCurrency.ts [hunk 18]
```diff
@@ -48,8 +49,8 @@ export function formatCurrency(
     currency: LiteralUnion<Keys | 'USD'> = 'USD',
     options?: FormatterCurrencyOptions,
 ): string {
-    const bn = new BigNumber(inputValue)
-    const { onlyRemainTwoDecimal = false } = options ?? {}
+    const { onlyRemainTwoDecimal = false, fiatCurrencyRate = 1 } = options ?? {}
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/helpers/formatCurrency.ts:53]: Keep test cases updating to date.

File: packages/shared-base/src/Messages/Events.ts [hunk 19]
```diff
@@ -256,6 +256,8 @@ export interface RequestsUpdatedEvent {
 
 export type WalletLockStatusUpdatedEvent = boolean
 
+export type walletLockTimeUpdatedEvent = void
```
Comment (Jack-Works) [inline packages/shared-base/src/Messages/Events.ts]: ? you don't have to do this

File: packages/web3-hooks/base/src/useFiatCurrencyRate.ts [hunk 20]
```diff
@@ -0,0 +1,9 @@
+import { FiatCurrencyRate } from '@masknet/web3-providers'
+
+import { useAsyncRetry } from 'react-use'
+import { useFiatCurrencyType } from './useFiatCurrencyType.js'
+
+export function useFiatCurrencyRate() {
+    const fiatCurrencyType = useFiatCurrencyType()
+    return useAsyncRetry(async () => FiatCurrencyRate.getRate(fiatCurrencyType), [fiatCurrencyType])
```
Comment (UncleBill) [inline packages/web3-hooks/base/src/useFiatCurrencyRate.ts]: Try react-query. I see you use this hook many times in this patch. With react-query's caching, it would be smoother and more efficient.

File: packages/web3-hooks/base/src/useWallet.ts [hunk 21]
```diff
@@ -18,5 +18,5 @@ export function useWallet<T extends NetworkPluginID>(
     const wallets = useWallets(pluginID, providerType)
     return useMemo(() => {
         return account ? wallets.find((x) => isSameAddress?.(x.address, account)) ?? null : null
-    }, [account, wallets.map((x) => x.address.toLowerCase()).join(',')])
+    }, [account, wallets.map((x) => x.address.toLowerCase() + x.name + x.owner).join(',')])
```
Comment (UncleBill) [inline packages/web3-hooks/base/src/useWallet.ts]: No need to serialize. `wallets` should be stable.

File: packages/mask/src/extension/popups/modals/ChooseCurrencyModal/index.tsx [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+            onClick={setFiatCurrencyType}>
+            <Box className={classes.itemBox}>
+                <FiatCurrencyIcon type={fiatCurrencyType} size={24} />
+                <Typography className={classes.text}>{resolveFiatCurrencyName(fiatCurrencyType)}</Typography>
+            </Box>
+            {selected ? (
+                <Icons.RadioButtonChecked size={20} />
+            ) : (
+                <Icons.RadioButtonUnChecked size={20} color={theme.palette.maskColor.line} />
+            )}
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/modals/ChooseCurrencyModal/index.tsx]: Just added a `RadioButton`, you can use it instead.

https://github.com/DimensionDev/Maskbook/blob/89b0f42e9f04dfe785d21ec0094bba61c01b6e27/packages/theme/src/Components/RadioButton/index.tsx#L1-L10

File: packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx (hunk 9 above)
Comment (hanszhoou) [reply packages/mask/src/extension/popups/modals/WalletAutoLockSettingModal/index.tsx:67]: This is not a timestamp. Just to require a number only input.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 16 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts]: `hasSecret` and `hasSafeSecret` is not for password set detection.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 17 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts]: It does not be treated as a password here.

File: packages/web3-shared/evm/src/libs/PayloadEditor.ts [hunk 23]
```diff
@@ -16,6 +15,7 @@ import {
     type EIP3085Descriptor,
     EthereumMethodType,
 } from '../types/index.js'
+import { abiCoder } from '../index.js'
```
Comment (guanbinrui) [inline packages/web3-shared/evm/src/libs/PayloadEditor.ts]: Use a specific relative path.

File: packages/web3-providers/src/Web3/Base/state/Settings.ts [hunk 24]
```diff
@@ -66,4 +66,8 @@ export class SettingsState implements Web3SettingsState {
             this.storage.nonFungibleAssetSourceType.initializedPromise,
         ]).then(() => {})
     }
+
+    async setDefaultCurrencyType(currencyType: CurrencyType) {
+        this.storage.currencyType.setValue(currencyType)
```
Comment (guanbinrui) [inline packages/web3-providers/src/Web3/Base/state/Settings.ts]:
```suggestion
        await this.storage.currencyType.setValue(currencyType)
```

File: packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx [hunk 25]
```diff
@@ -147,32 +192,50 @@ export default function ChangeOwner() {
...
+    const personaManager = useMemo(
+        () => personaManagers?.find((x) => isSameAddress(wallet?.owner, x.address)),
+        [personaManagers, wallet],
+    )
+
+    const managerAddress = walletManager?.address ?? personaManager?.address
+
+    const { value: gas } = useAsync(async () => {
+        const contract = Contract.getWeb3Contract<Wallet>(wallet?.address, WalletABI as AbiItem[])
+        if (!manageAccount?.address) return
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx]:
```suggestion
        if (!manageAccount?.address || !wallet?.address) return
```

File: packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx [hunk 26]
```diff
@@ -184,211 +247,212 @@ export default function ChangeOwner() {
...
+                            <Typography className={classes.primaryItemText}>{wallet?.name}</Typography>
+                            <Typography className={classes.primaryItemSecondText}>
+                                <span className={classes.walletAddress}>{wallet?.address}</span>
+                                {wallet?.address ? (
+                                    <>
+                                        <CopyButton size={16} className={classes.linkIcon} text={wallet.address} />
+                                        <Link
+                                            href={ExplorerResolver.addressLink(chainId, wallet.address)}
+                                            target="_blank"
+                                            title="View on Explorer"
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx]: i18n

File: packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx [hunk 27]
```diff
@@ -184,211 +247,212 @@ export default function ChangeOwner() {
...
+                                    <>
+                                        <CopyButton size={16} className={classes.linkIcon} text={managerAddress} />
+                                        <Link
+                                            href={
+                                                walletManager
+                                                    ? ExplorerResolver.addressLink(chainId, managerAddress)
+                                                    : urlcat('https://web3.bio/', { s: managerAddress })
+                                            }
+                                            target="_blank"
+                                            title="View on Explorer"
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx]: Same here.

File: packages/mask/src/plugins/WalletService/services/wallet/locker.ts [hunk 28]
```diff
@@ -25,15 +26,16 @@ export async function unlockWallet(unverifiedPassword: string) {
     }
 }
 
-const DEFAULT_LOCK_TIME = 1000 * 60 * 15 // 15 mins
-
 let autoLockTimer: NodeJS.Timeout | undefined
 
-export function setAutoLockTimer() {
+export async function setAutoLockTimer() {
+    const autoLockTime = await getAutoLockerTime()
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/locker.ts]:
```suggestion
    const autoLockDuration = await getAutoLockerDuration()
```

File: packages/mask/src/plugins/WalletService/services/wallet/locker.ts [hunk 29]
```diff
@@ -25,15 +26,16 @@ export async function unlockWallet(unverifiedPassword: string) {
...
+
     clearTimeout(autoLockTimer)
+
+    if (autoLockTime === 0) return
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/locker.ts]: Add detection of negative values.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 30]
```diff
@@ -215,7 +215,14 @@ export async function exportPrivateKey(address: string, unverifiedPassword?: str
 
 export async function exportKeyStoreJSON(address: string, unverifiedPassword?: string) {
     if (unverifiedPassword) await password.verifyPasswordRequired(unverifiedPassword)
-    const password_ = await password.INTERNAL_getPasswordRequired()
+    let password_: string = ''
+    try {
+        password_ = await password.INTERNAL_getPasswordRequired()
+    } catch {
+        if (!unverifiedPassword) throw new Error('No password set yet or expired.')
+        password_ = await database.decryptSecret(unverifiedPassword)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts]: Not necessary here. Line #217 has verified the password.

File: packages/mask/src/plugins/WalletService/services/wallet/password.ts [hunk 31]
```diff
@@ -61,7 +75,8 @@ export async function verifyPasswordRequired(unverifiedPassword: string) {
 
 export async function changePassword(oldPassword: string, newPassword: string) {
     validatePasswordRequired(newPassword)
-    validatePasswordRequired(oldPassword)
+    const isVerify = await verifyPassword(oldPassword)
+    if (!isVerify) throw new Error('Incorrect payment password.')
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/password.ts]:
```suggestion
    const verified = await verifyPassword(oldPassword)
    if (!verified) throw new Error('Incorrect payment password.')
```

Comment (guanbinrui) [reply packages/mask/src/plugins/WalletService/services/wallet/password.ts]: Did you mean `verifyPassword(newPassword)`?

Comment (hanszhoou) [reply packages/mask/src/plugins/WalletService/services/wallet/password.ts]: No.

File: packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts [hunk 32]
```diff
@@ -0,0 +1,26 @@
+import { CrossIsolationMessages } from '@masknet/shared-base'
+import { PluginDB } from '../../../database/Plugin.db.js'
+
+const DEFAULT_LOCK_TIME = 1000 * 60 * 15 // 15 mins
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts]: Rename it to `DEFAULT_LOCK_DURATION`.

File: packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+
+const ID = 'unlock-time'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts]:
```suggestion
const ID = 'locker'
```

File: packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts [hunk 34]
```diff
@@ -0,0 +1,26 @@
...
+
+export type LockerRecord = {
+    type: 'unlock-time'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts]:
```suggestion
    type: 'locker'
```

File: packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts [hunk 35]
```diff
@@ -0,0 +1,26 @@
...
+    id: string
+    time: number // ms
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/database/locker.ts]:
```suggestion
    duration: number // ms
```

File: packages/mask/src/plugins/WalletService/services/wallet/password.ts [hunk 36]
```diff
@@ -61,7 +75,8 @@ export async function verifyPasswordRequired(unverifiedPassword: string) {
 
 export async function changePassword(oldPassword: string, newPassword: string) {
     validatePasswordRequired(newPassword)
-    validatePasswordRequired(oldPassword)
+    const verified = await verifyPassword(oldPassword)
+    if (!verified) throw new Error('Incorrect payment password.')
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/password.ts]:
```suggestion
    await verifyPasswordRequired(oldPassword)
```

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 37]
```diff
@@ -192,8 +192,8 @@ export async function exportMnemonicWords(address: string, unverifiedPassword?:
 }
 
 export async function exportPrivateKey(address: string, unverifiedPassword?: string) {
-    if (unverifiedPassword) await password.verifyPasswordRequired(unverifiedPassword)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts:195]: Please revert this change.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 38]
```diff
@@ -214,8 +214,8 @@ export async function exportPrivateKey(address: string, unverifiedPassword?: str
 }
 
 export async function exportKeyStoreJSON(address: string, unverifiedPassword?: string) {
-    if (unverifiedPassword) await password.verifyPasswordRequired(unverifiedPassword)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts:217]: Please revert this change.

**Review by guanbinrui — APPROVED** [review]
